src/object_async/hello_async.cpp

```diff
@@ -0,0 +1,227 @@
+#include "hello_async.hpp"
+
+#include <exception>
+#include <stdexcept>
+#include <iostream>
+#include <map>
+
+// If this was not defined within a namespace, it would be in the global scope.
+namespace object_async {
+
+  // Custom constructor, assigns custom name passed in from Javascript world.
+  // This constructor uses member init list via the semicolon, aka "direct initialization" 
+  // which is more efficient than using assignment operators.
+  HelloObjectAsync::HelloObjectAsync(std::string name) : 
+    name_(name) {}
+
+  // Triggered from Javascript world when calling "new HelloObjectAsync()" or "new HelloObjectAsync(name)"
+  NAN_METHOD(HelloObjectAsync::New) {
+    if (info.IsConstructCall())
+    {
+        try
+        {   
+            if (info.Length() >= 1) {
+              if (info[0]->IsString()) 
+              {
+                std::string name = *v8::String::Utf8Value(info[0]->ToString());
+                auto *const self = new HelloObjectAsync(name);
+                self->Wrap(info.This());
+              }
+              else
+              {
+                return Nan::ThrowTypeError(
+                    "arg must be a string");
+              }
+            }
+            else {
+                auto *const self = new HelloObjectAsync();
+                self->Wrap(info.This());
+            } 
+
+        }
+        catch (const std::exception &ex)
+        {
+            return Nan::ThrowTypeError(ex.what());
+        }
+
+        info.GetReturnValue().Set(info.This());
+    }
+    else
+    {
+        return Nan::ThrowTypeError(
+            "Cannot call constructor as function, you need to use 'new' keyword");
+    }
```

Comment:
@daniel-j-h - really great catch 🙇 . I agree that handling the case where a user forgets to use `new` in Javascript is important.  I'm aware that some JS developers like to hide the complexity and allow object creation to work without `new`. But my preference would be to throw since this is a consistent pattern that has worked in other node C++ modules I have written. And I've never seen a complaint about this.